src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,180 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size = 0;
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            head = tail = new Node(null, value, null);
+        } else {
+            Node notTheTailAnyMore = tail;
+            tail = new Node(notTheTailAnyMore, value, null);
+            notTheTailAnyMore.next = tail;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("given index does not exist");
+        } else {
+            if (index == size) {
+                add(value);
+            } else if (index == 0) {
+                Node notTheHeadAnyMore = head;
+                head = new Node(null, value, notTheHeadAnyMore);
+                notTheHeadAnyMore.prev = head;
+                size++;
+            } else {
+                int currentIndex = 0;
+                if (size > 1 && index > (int) (size / 2)) {
+                    currentIndex = (int) (size / 2) - 1;
+                }
+                Node currentNode = head;
+                while (currentIndex != index) {
+                    currentNode = currentNode.next;
+                    currentIndex++;
+                }
+                Node newNode = new Node(currentNode.prev, value, currentNode);
+                currentNode.prev.next = newNode;
+                currentNode.prev = newNode;
+
+                size++;
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
+    }
+
+    public Node getNode(int index) {
```

Comment:
Common mistake: Don't forget about access modifiers. In this case, the `getNode` method should be `private` because it's not part of the public API and should not be accessible from outside the class.